### PR TITLE
feat: add support for types from ascii crate

### PR DESCRIPTION
**Pull request by mina86**

### CODE REVIEW

File: borsh/src/de/mod.rs [hunk 1]
```diff
@@ -370,6 +370,28 @@ impl BorshDeserialize for String {
...
+#[test]
+fn test_ascii() {
+    let encoded = borsh::to_vec("foo").unwrap();
+
+    let got = ascii::AsciiString::deserialize_reader(&mut &encoded[..]).unwrap();
+    assert_eq!("foo", got.as_str());
+
+    let encoded = borsh::to_vec("żółw").unwrap();
+    ascii::AsciiString::deserialize_reader(&mut &encoded[..]).unwrap_err();
+}
```
Comment (dj8yfo) [inline borsh/src/de/mod.rs:393]: please move all added tests to integration tests `borsh/tests/test_ascii_strings.rs`

File: borsh/src/ser/mod.rs [hunk 2]
```diff
@@ -210,6 +210,32 @@ impl BorshSerialize for String {
...
+
+#[cfg(feature = "ascii")]
+#[test]
+fn test_ascii() {
+    let val = ascii::AsciiStr::from_ascii("foo").unwrap();
+    let encoded = borsh::to_vec(val).unwrap();
+
+    let got: String = borsh::BorshDeserialize::deserialize_reader(&mut &encoded[..]).unwrap();
+    assert_eq!("foo", got);
+}
```
Comment (dj8yfo) [inline borsh/src/ser/mod.rs:237]: please add 2 lines to `.github/test.sh` to actually test this new feature (`std` and `no-std`),
similar to https://github.com/near/borsh-rs/blob/master/.github/test.sh#L10 and https://github.com/near/borsh-rs/blob/master/.github/test.sh#L16

File: borsh/src/schema.rs [hunk 3]
```diff
@@ -384,6 +384,19 @@ impl BorshSchema for String {
...
+
+#[cfg(feature = "ascii")]
+impl BorshSchema for ascii::AsciiString {
+    #[inline]
+    fn add_definitions_recursively(definitions: &mut BTreeMap<Declaration, Definition>) {
+        str::add_definitions_recursively(definitions);
+    }
+    #[inline]
+    fn declaration() -> Declaration {
+        str::declaration()
```
Comment (dj8yfo) [inline borsh/src/schema.rs:396]: please change declaration to `AsciiString` in both cases

File: borsh/src/schema.rs [hunk 4]
```diff
@@ -401,6 +414,18 @@ impl BorshSchema for str {
...
 
+#[cfg(feature = "ascii")]
+impl BorshSchema for ascii::AsciiStr {
+    #[inline]
+    fn add_definitions_recursively(definitions: &mut BTreeMap<Declaration, Definition>) {
+        str::add_definitions_recursively(definitions);
+    }
+    #[inline]
+    fn declaration() -> Declaration {
+        str::declaration()
```
Comment (dj8yfo) [inline borsh/src/schema.rs:425]: -> `AsciiString`

**Review by dj8yfo — CHANGES_REQUESTED** [review]

File: borsh/src/de/mod.rs [hunk 5]
```diff
@@ -370,6 +370,28 @@ impl BorshDeserialize for String {
...
+#[cfg(feature = "ascii")]
+#[test]
+fn test_ascii() {
+    let encoded = borsh::to_vec("foo").unwrap();
+
+    let got = ascii::AsciiString::deserialize_reader(&mut &encoded[..]).unwrap();
+    assert_eq!("foo", got.as_str());
+
+    let encoded = borsh::to_vec("żółw").unwrap();
+    ascii::AsciiString::deserialize_reader(&mut &encoded[..]).unwrap_err();
```
Comment (dj8yfo) [inline borsh/src/de/mod.rs:392]: move this test to `borsh/test/test_de_errors.rs` with explicit message check

File: borsh/src/de/mod.rs [hunk 6]
```diff
@@ -370,6 +370,28 @@ impl BorshDeserialize for String {
...
+            .map_err(|err| Error::new(ErrorKind::InvalidData, err.to_string()))
+    }
+}
+
+#[cfg(feature = "ascii")]
+#[test]
+fn test_ascii() {
+    let encoded = borsh::to_vec("foo").unwrap();
+
+    let got = ascii::AsciiString::deserialize_reader(&mut &encoded[..]).unwrap();
```
Comment (dj8yfo) [inline borsh/src/de/mod.rs:388]: please add capturing a snapshot of byte array with `insta`

File: borsh/src/de/mod.rs [hunk 7]
```diff
@@ -370,6 +370,28 @@ impl BorshDeserialize for String {
...
+    fn deserialize_reader<R: Read>(reader: &mut R) -> Result<Self> {
+        let bytes = Vec::<u8>::deserialize_reader(reader)?;
+        ascii::AsciiString::from_ascii(bytes)
+            .map_err(|err| Error::new(ErrorKind::InvalidData, err.to_string()))
+    }
+}
+
+#[cfg(feature = "ascii")]
+#[test]
+fn test_ascii() {
```
Comment (dj8yfo) [inline borsh/src/de/mod.rs:385]: please change logic and naming of tests:
`test_ascii_string_to_ascii_string_roundtrip`
`test_ascii_str_to_ascii_string_roundtrip`
And 2 separate tests for `ascii`/`std`-`Strings` interoperability (which are optional):
`test_str_to_ascii_string_maybe_roundtrip`
`test_ascii_str_to_string_roundtrip`.

**Review by dj8yfo — CHANGES_REQUESTED** [review]

File: borsh/tests/test_strings.rs [hunk 8]
```diff
@@ -2,33 +2,80 @@
...
+#[test]
+fn test_ascii_char() {
+    use ascii::AsciiChar;
+    use ascii::AsciiChar::Dot;
+
+    let buf = to_vec(&Dot).unwrap();
+    assert_eq!(".".as_bytes(), buf);
+    assert_eq!(Dot, from_slice::<AsciiChar>(&buf).unwrap());
+
+    from_slice::<AsciiChar>(&[b'\x80']).unwrap_err();
```
Comment (dj8yfo) [inline borsh/tests/test_strings.rs:80]: please move to `borsh/test/test_de_errors.rs`

File: borsh/tests/test_strings.rs [hunk 9]
```diff
@@ -2,33 +2,80 @@
...
+test_string!(non_ascii, "💩", true);
+
+#[test]
+fn test_non_utf8() {
+    let data: [u8; 4] = [0xbf, 0xf3, 0xb3, 0x77];
+    let buf = to_vec(&data[..]).unwrap();
+    from_slice::<String>(&buf).unwrap_err();
+    #[cfg(feature = "ascii")]
+    from_slice::<ascii::AsciiString>(&buf).unwrap_err();
+}
```
Comment (dj8yfo) [inline borsh/tests/test_strings.rs:68]: please move to `borsh/test/test_de_errors.rs` as 2 separate tests with explicit err message check

File: borsh/src/schema.rs [hunk 10]
```diff
@@ -401,6 +402,48 @@ impl BorshSchema for str {
...
+impl BorshSchema for ascii::AsciiChar {
+    #[inline]
+    fn add_definitions_recursively(definitions: &mut BTreeMap<Declaration, Definition>) {
+        add_definition(Self::declaration(), Definition::Primitive(1), definitions);
+    }
+    #[inline]
+    fn declaration() -> Declaration {
+        "AsciiChar".into()
+    }
+}
```
Comment (dj8yfo) [inline borsh/src/schema.rs:445]: please add tests for schema generated for these 3 types right after https://github.com/near/borsh-rs/blob/master/borsh/src/schema.rs#L1025-L1058

File: .github/test.sh [hunk 11]
```diff
@@ -5,16 +5,18 @@ export INSTA_UPDATE=no
 pushd borsh
 cargo test --no-run
 cargo test
+cargo test --features ascii
```
Comment (dj8yfo) [inline .github/test.sh:8]:
```suggestion
cargo test --features ascii,unstable__schema
```

File: .github/test.sh [hunk 12]
```diff
@@ -5,16 +5,18 @@ export INSTA_UPDATE=no
...
 cargo test --features derive
 cargo test --features unstable__schema
 cargo test --test test_rc --features rc
 cargo test --test test_hash_map --test test_btree_map --features de_strict_order
 
 cargo test --no-default-features
+cargo test --no-default-features --features ascii
```
Comment (dj8yfo) [inline .github/test.sh:15]:
```suggestion
cargo test --no-default-features --features ascii,unstable__schema
```

**Review by dj8yfo — CHANGES_REQUESTED** [review]

File: borsh/src/de/mod.rs (hunk 7 above)
Comment (mina86) [reply borsh/src/de/mod.rs]: I will definitely not do that.

File: borsh/tests/test_strings.rs [hunk 13]
```diff
@@ -2,33 +2,80 @@
...
-test_string!(test_empty_string, "", true);
-test_string!(test_a, "a", true);
-test_string!(test_hello_world, "hello world", true);
-test_string!(test_x_1024, "x".repeat(1024), true);
-test_string!(test_x_4096, "x".repeat(4096), false);
-test_string!(test_x_65535, "x".repeat(65535), false);
-test_string!(test_hello_10, "hello world!".repeat(30), true);
-test_string!(test_hello_1000, "hello world!".repeat(1000), false);
-test_string!(test_non_ascii, "💩", true);
+test_string!(empty_string, "", true);
```
[the remark this answers is not in the document]
Comment (mina86) [reply borsh/tests/test_strings.rs:51]: > please move all ascii tests to borsh/tests/test_ascii_strings.rs

How is duplicating the same test_string/test_ascii_string invocations in two different files better than keeping it DRY in a single file? You’re asking me to duplicate code for no reason.

**Comment by mina86** [comment]
Address all but moving stuff to separate file and duplicating test cases.  I’ve no intention of doing that ’cuz it’s silly.

**Review by dj8yfo — APPROVED** [review]

**Pull request merged**